### PR TITLE
ossl tls: Added support to configure certificate verify depth

### PR DESCRIPTION
**Pull request by alorbach**
Support added in omfwd as instance parameter:
streamdriver.TlsVerifyDepth
Support added in imtcp as module parameter:
streamdriver.TlsVerifyDepth

Can be 2 or higher.
Support added into ossl driver

Parameter
streamdriver.TlsVerifyDepth

closes: https://github.com/rsyslog/rsyslog/issues/4035

### CODE REVIEW

File: runtime/tcpsrv.c [hunk 1]
```diff
@@ -1000,6 +1000,7 @@ BEGINobjConstruct(tcpsrv) /* be sure to specify the object type also in END macr
 	pThis->bUseFlowControl = 1;
 	pThis->pszDrvrName = NULL;
 	pThis->bPreserveCase = 1; /* preserve case in fromhost; default to true. */
+	pThis->DrvrTlsVerifyDepth = 25;
```
Comment (rgerhards) [inline runtime/tcpsrv.c]: This looks inconsistent to the other default. I would suggest to leave the default at 0, as this avoids the API call and so the library default is used (whatever it is). I think this is the least confusing option for users.

File: tools/omfwd.c [hunk 2]
```diff
@@ -85,6 +85,8 @@ typedef struct _instanceData {
 	int iStrmDrvrMode;
 	int iStrmDrvrExtendedCertCheck; /* verify also purpose OID in certificate extended field */
 	int iStrmDrvrSANPreference; /* ignore CN when any SAN set */
+#	define OMFWD_VERIFY_DEPTH 100
```
Comment (rgerhards) [inline tools/omfwd.c]: 0 makes a better default (see other comment)

File: plugins/imtcp/imtcp.c [hunk 3]
```diff
@@ -133,6 +134,8 @@ struct modConfData_s {
 	int iStrmDrvrMode; /* mode for stream driver, driver-dependent (0 mostly means plain tcp) */
 	int iStrmDrvrExtendedCertCheck; /* verify also purpose OID in certificate extended field */
 	int iStrmDrvrSANPreference; /* ignore CN when any SAN set */
+#	define IMTCP_VERIFY_DEPTH 100
```
Comment (rgerhards) [inline plugins/imtcp/imtcp.c]: 0 makes a better default (see other comment)

Comment (alorbach) [reply plugins/imtcp/imtcp.c]: yes this can be removed now, as default is handeled in nsd_ossl.c now

File: runtime/nsd_ossl.c [hunk 4]
```diff
@@ -78,6 +78,7 @@ DEFobjCurrIf(nsd_ptcp)
 	#define RSYSLOG_BIO_number_read(SSLBIO) SSLBIO->num
 	#define RSYSLOG_BIO_number_written(SSLBIO) SSLBIO->num
 #endif
+#define OSSL_VERIFY_DEPTH 100
```
Comment (rgerhards) [inline runtime/nsd_ossl.c]: I guess this can be removed?

Comment (alorbach) [reply runtime/nsd_ossl.c]: We need the default here as we set the verify_depth default here for the existing API call which was hardcoded to 2 before.

Comment (alorbach) [reply runtime/nsd_ossl.c]: Removed need of OSSL_VERIFY_DEPTH , it should be fine now

**Comment by rgerhards** [comment]
doc: https://github.com/rsyslog/rsyslog-doc/pull/878

File: tests/imtcp-tls-ossl-basic-verifydepth.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+global(	defaultNetstreamDriverCAFile="'$srcdir/tls-certs/ca.pem'"
+	defaultNetstreamDriverCertFile="'$srcdir/tls-certs/cert.pem'"
+	defaultNetstreamDriverKeyFile="'$srcdir/tls-certs/key.pem'"
+)
+module(	load="../plugins/imtcp/.libs/imtcp"
+	StreamDriver.Name="ossl"
+	StreamDriver.TlsVerifyDepth="5" 
+	StreamDriver.Mode="1"
+	StreamDriver.AuthMode="x509/certvalid" )
+input(	type="imtcp"
```
Comment (rgerhards) [inline tests/imtcp-tls-ossl-basic-verifydepth.sh]: Please update the test plumbing - this is the old, racy version.  I have just fixed similar tests here: https://github.com/rsyslog/rsyslog/pull/4102/files

Pls also have all input parameters on one line - this simplifies future changes via `sed`. If there are very many parameters, at least the `type` and `port` should be in one line and in this order.

Comment (alorbach) [reply tests/imtcp-tls-ossl-basic-verifydepth.sh]: done

**Comment by rgerhards** [comment]
Note to myself: fix test after merge

**Comment by alorbach** [comment]
> I'll do the fix after merge

Ok thanks

**Pull request merged**